DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
+    "    with graph.as_default():\n",
+    "        if model_path is None:\n",
+    "            # Init all variables for first use\n",
+    "            session.run(tf.initialize_all_variables())\n",
+    "        else:\n",
+    "            # Loading the model into the session\n",
+    "            execution_context[3].restore(session, model_path)\n",
+    "\n",
+    "        # Start queue runner threads\n",
+    "        runner_threads = tf.train.start_queue_runners(sess=session)\n",
     "\n",
-    "    # Average tower gradients\n",
-    "    avg_tower_gradients = average_gradients(tower_gradients)\n",
+    "        # Start importer's queue threads\n",
+    "        queue_threads = execution_context[1].start_queue_threads(session)\n",
     "\n",
-    "    # Add logging of averaged gradients\n",
-    "    log_grads_and_vars(avg_tower_gradients)\n",
+    "    return session"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following helper method stops an execution contect. \n",
+    "Before closing the provided `session`, it will persist the contained model to disk. \n",
+    "The model's filename is constructed from `checkpoint_path` and `global_step` and will be returned."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def stop_execution_context(execution_context, session, checkpoint_path=None, global_step=None):\n",
+    "    \n",
+    "    # If the model is not persisted, we'll return 'None'\n",
+    "    hibernation_path = None\n",
+    "    \n",
+    "    if checkpoint_path is not None and global_step is not None:\n",
+    "        # Saving session's model into checkpoint dir\n",
+    "        hibernation_path = execution_context[3].save(session, checkpoint_path, global_step=global_step)\n",
+    "        \n",
+    "    # Free all allocated resources of the session\n",
+    "    session.close()\n",
+    "    \n",
+    "    return hibernation_path"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's introduce the main routine for training and inference. \n",
+    "It takes a started execution context (given by `execution_context` and `session`), an optional `epoch` index and a flag wheather to calculate WER report data (`query_report`) or not.\n",
+    "Its main duty is to iterate over all batches and calculate the mean loss.\n",
+    "If an epoch is provided, it will also do back propagation - aka training.\n",
+    "It will return a batch set result tuple which consists of the mean loss and - if asked for report data - the mean accuracy and individual results tuple."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def apply_batches(execution_context, session, epoch=-1, query_report=False):\n",
```

Comment:
Why is this called `apply_batches`? I think a name that's more descriptive of its function would be good. Something like `calculate_loss` (and splitting out the reporting would make this a more accurate name), or assuming the split isn't feasible, `calculate_loss_and_wer_report` or something?